### DISCUSSION TITLE
Creating an alias to a `case class`

### DISCUSSION THREAD

**Discussion by goekce** [SpinalHDL/SpinalHDL#1596] [Q&A] [answered]
I want to create a project template for simple projects that uses less boilerplate code. In the following code, I put the design and simulation together and would like to create an alias for the `Multiplier`, so I have to change the identifier `Multiplier` if I want to rename it. 

```scala
import spinal.core._

object Design extends Multiplier()

case class Multiplier() extends Component {
  val io = new Bundle {
    val a = in port UInt(4 bits)
    val b = in port UInt(4 bits)
    val y = out port UInt(8 bits)
  }
  val y = Reg(UInt(8 bits))
  y := io.a * io.b
  io.y := y
}

object Gen extends App {
  Config.spinal.generateVerilog(Design)  ///////
}

import spinal.core.sim._

object Sim extends App {
  Config.sim.compile(Design).doSim { dut =>  ///////
    dut.clockDomain.forkStimulus(period = 10)

    for (idx <- 0 to 99) {
      dut.io.a.randomize()
      dut.io.b.randomize()

      dut.clockDomain.waitRisingEdge()
    }
  }
}
```
However, I don't want to instantiate the `Multiplier` 😟.

As I understand, only `object` and `class` are allowed at the top level, so I could not find a solution other than `object Design extends Multiplier()`, but this approach already instantiates `Multiplier`. Is aliasing without instantiation possible?

What I tried
- `class Design() extends Multiplier` (and using `Design()`
- `case class Design() extends Multiplier`
both in addition to Config.spinal.generateVerilog(Design())`.

**Comment by goekce** [comment, marked as answer]
I found a solution:

```scala
object Design {
  def apply() = new Multiplier()
}
```
😌

Apparently, I have to use an uninstantiated reference and the companion object method `apply`.

**Comment by Dolu1990** [comment]
Hi,
```scala
object Design extends Multiplier()
```
^ doesnt work because object will be created as a singleton/static thing.

Your solution is good.
Else you can : 
```scala
class Design extends Multiplier
class Multiplier() extends Component { ... }  // remove case here
...
Config.spinal.generateVerilog(new Design)
```

**Reply by goekce** [reply]
I read in the docs that [`case class` should be preferred](https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Getting%20Started/Scala%20Guide/coding_conventions.html#class-vs-case-class) because it has a shorter syntax and other useful methods. So this is an exception to this rule, right?

**Reply by Dolu1990** [reply]
Right and right.
Overall, Scala 2 as some flaws, and this one has been fixed in scala 3
